trackintel/preprocessing/trips.py

```diff
@@ -0,0 +1,331 @@
+                # option 1: no gaps allowed - start search again
+                if max_nr_gaps == 0:
+                    start_candidates = [row.name]
+                    continue
+                # option 2: gaps allowed - search further
+                else:
+                    start_candidates.append(np.nan)
+
+        # Add this point as a candidate
+        start_candidates.append(row.name)
+
+        # Check whether endpoint would be an unkown activity
+        if pd.isna(row["destination_staypoint_id"]):
+            continue
+
+        # keep a list of which candidates to remove (because of time frame)
+        new_list_start = 0
+
+        # check distance to all candidates (except the ones that are too close)
+        for j, cand in enumerate(start_candidates):
+            # gap
+            if np.isnan(cand):
+                continue
+
+            # check time difference - if time too long, we can remove the candidate
+            cand_start_time = user_trip_df.loc[cand, "started_at"]
+            if end_time - cand_start_time > max_time:
+                new_list_start = j + 1
+                continue
+
+            # check whether the start-end candidate of a tour is an unkown activity
+            if pd.isna(user_trip_df.loc[cand, "origin_staypoint_id"]):
+                continue
+
+            # TODO: compute length of triplegs and sum - must be larger than minthresh
+            # check if endpoint of trip = start location of cand
+            if stps_w_locs is not None:
+                end_start_at_same_loc = _check_same_loc(
+                    user_trip_df.loc[cand, "origin_staypoint_id"],  # start stp of first trip
+                    row["destination_staypoint_id"],  # destination stp of current trip
+                    stps_w_locs,
+                )
+            else:
+                # check distance between point 1: end point of current trip, point 2: start of first trip on tour
+                end_start_at_same_loc = _check_max_dist(row[geom_col][1], user_trip_df.loc[cand, geom_col][0], max_dist)
+
+            # print("Check distance to start", cand, end_point, cand_start_point, point_dist)
+            if end_start_at_same_loc:
+                # Tour found!
+                # collect the trips on the tour in a list
+                non_gap_trip_idxs = [c for c in start_candidates[j:] if ~np.isnan(c)]
+                tour_candidate = user_trip_df[user_trip_df.index.isin(non_gap_trip_idxs)]
+                tours.append(_create_tour_from_stack(tour_candidate, stps_w_locs, max_dist, max_time, geom_col))
+
+                nr_gaps = np.sum(np.isnan(np.array(start_candidates[j:])))
+
+                if nr_gaps > max_nr_gaps:
+                    # No tour found, too many gaps inbetween
+                    continue
+                # print("Tour found!", tour_candidate.head())
+
+                # remove trips that were on the tour
+                start_candidates = start_candidates[:j]
```

Comment:
I am not sure if it is safe to change the list you are iterating on during the loop. I would rather not do this. 

I am not sure if you are potentially losing subtours by cutting of the start. Consider this example: 
Home ---trip1--->Work---trip2--->Lunch---trip3--->Work---trip4--->Home

So if you going trip by trip, the first tour are are detecting is trip2 + trip3. Then you would delete trip2 and trip3 of `start_candidates` and the next tour would be only trip1+trip4. I am actually not entirely sure which solution we should prefer but it might be that the big tour should contain all trips performed from home to home.